superduperdb/db/mongodb/cdc/handler.py

```diff
@@ -35,120 +34,54 @@ def __init__(self, db: DB, stop_event: threading.Event):
-                job=FunctionJob(
-                    callable=task_callable,
-                    args=[indexing_listener_identifier, serialized_cdc_query, ids],
-                    kwargs={},
-                ),
-            )
-        if task != 'delete':
-            assert indexing_listener_identifier
-            model, _, key = indexing_listener_identifier.rpartition('/')
-            task_workflow.add_edge(
-                f'{model}.predict({key})',
-                f'{task_name}({indexing_listener_identifier})',
-            )
-        return task_workflow
-
-    def on_create(self, packet: Packet) -> None:
-        ids = packet.ids
-        cdc_query = packet.query
-        self.submit_task_workflow(cdc_query=cdc_query, ids=ids)
-
-    def on_update(self, packet: Packet) -> None:
-        self.on_create(packet)
-
-    def on_delete(self, packet: Packet) -> None:
-        ids = packet.ids
-        cdc_query = packet.query
-        self.submit_task_workflow(cdc_query=cdc_query, ids=ids, task="delete")
-
-    def _handle(self, packet: Packet) -> None:
-        if packet.event_type == DBEvent.insert:
-            self.on_create(packet)
-        elif packet.event_type == DBEvent.update:
-            self.on_update(packet)
-        elif packet.event_type == DBEvent.delete:
-            self.on_delete(packet)
-
-    def get_batch_from_queue(self):
-        """
-        Get a batch of packets from task queue, with a timeout.
-        """
-        packets = []
-        try:
-            for _ in range(self._QUEUE_BATCH_SIZE):
-                packets.append(cdc_queue.get(block=True, timeout=self._QUEUE_TIMEOUT))
-                if self._stop_event.is_set():
-                    return 0
+        vi = db.load(identifier=identifier, type_id='vector_index')
+        assert isinstance(vi, VectorIndex)
 
-        except queue.Empty:
-            if len(packets) == 0:
-                return None
-        return Packet.collate(packets)
+        assert not isinstance(vi.indexing_listener, str)
+        listener_id = vi.indexing_listener.identifier
 
-    def run(self):
-        while not self._stop_event.is_set():
-            try:
-                packets = self.get_batch_from_queue()
-                if packets:
-                    self._handle(packets)
-                if packets == 0:
-                    break
-            except Exception as exc:
-                traceback.print_exc()
-                logging.info(f'Error while handling cdc batches :: reason {exc}')
+        args = [listener_id, serialized_query, packet.ids]
+        job = FunctionJob(callable=task_callable, args=args)
+        workflow.add_node(f'{task_name}({listener_id})', job=job)
+
+        return listener_id
+
+    listener_ids = [add_node(i) for i in db.show('vector_index')]
+    if not packet.is_delete:
+        assert listener_ids
+        listener_id = listener_ids[-1]
+        model, _, key = listener_id.rpartition('/')
```

Comment:
I think we have used

rpartition('/') a lot of time around this project, so may be we can create a 
SuperduperID() class to represent ids with model/key things, 

and then have a function like

```python
def unpack():
    return self._id.rpartion('/')
``` 
@blythed

Reply:
Oh, that's a promising idea, to extract that model/key code!

I want to note here that "informal encodings" like this, where you have hand-written code to encode and decode some data structure to and from a string, are a proven source of bugs in code over decades. 🥇 

(And in fact it already resulted in one issue here.)

Given we have this, putting it into one place is logical.

One possibility is:

```
class ModelKey(str):
    @cached_property
    def model(self):
        return self.rpartion('/')[0]

    @cached_property
    def key(self):
        return self.rpartion('/')[1]
```

another

```
class ModelKey(str):
    def __init__(self, model: str, key: t.Optional[str] = None):
        if key is None:
            self.model, _, self.key = model.rpartion('/')
        else:
            self.model, self.key = model, key

    def __str__(self):
        return f'{self.model}/{self.key}'
```

Reply:
yes something like this!